plugins/hlint-hls-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,701 @@
+import Control.Monad.Extra
+import Control.Monad.Trans.Maybe
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except
+import qualified Data.Aeson as Aeson
+import Data.Aeson.Types (ToJSON(..), FromJSON(..), Value(..), Result(..))
+import Data.Binary
+import qualified Data.ByteString as BS
+import Data.Either.Extra
+import Data.Foldable
+import Data.Functor
+import qualified Data.HashMap.Strict as Map
+import qualified Data.HashSet as HashSet
+import Data.Hashable
+import Data.List
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+import qualified Data.Set as Set
+import Data.Set (Set)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Typeable
+import Data.Typeable (Typeable)
+import Development.IDE.Core.OfInterest
+import Development.IDE.Core.RuleTypes
+import Development.IDE.Core.Rules
+import Development.IDE.Core.Service
+import Development.IDE.Core.Shake
+import Development.IDE.GHC.Util (hscEnv)
+import Development.IDE.LSP.Server
+import Development.IDE.Plugin
+import Development.IDE.Types.Diagnostics as D
+import Development.IDE.Types.Location
+import Development.IDE.Types.Logger
+import Development.Shake
+-- import Development.Shake hiding ( Diagnostic )
+import GHC hiding (DynFlags(..))
+import GHC.Generics
+import GHC.Generics (Generic)
+import SrcLoc
+import HscTypes (ModIface, ModSummary)
+
+#ifndef GHC_LIB
+import GHC (DynFlags(..))
+import HscTypes (hsc_dflags)
+#else
+import RealGHC (DynFlags(..))
+import RealGHC.HscTypes (hsc_dflags)
+import qualified RealGHC.EnumSet as EnumSet
+import Language.Haskell.GhclibParserEx.GHC.Driver.Session as GhclibParserEx (readExtension)
+#endif
+
+import Ide.Logger
+import Ide.Types
+import Ide.Plugin
+import Ide.PluginUtils
+import Language.Haskell.HLint
+import Language.Haskell.HLint as Hlint
+import qualified Language.Haskell.LSP.Core as LSP
+import Language.Haskell.LSP.Messages
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types      as LSP
+import qualified Language.Haskell.LSP.Types.Lens as LSP
+import System.Directory
+import System.Directory.Extra as Dir
+import System.Environment.Blank
+import System.FilePath
+import System.IO.Error
+import Text.Regex.TDFA.Text()
+
+-- ---------------------------------------------------------------------
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId = (defaultPluginDescriptor plId)
+  { pluginRules = rules
+  , pluginCommands =
+      [ PluginCommand "applyOne" "Apply a single hint" applyOneCmd
+      , PluginCommand "applyAll" "Apply all hints to the file" applyAllCmd
+      ]
+    , pluginCodeActionProvider = Just codeActionProvider
+  }
+
+data GetHlintDiagnostics = GetHlintDiagnostics
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintDiagnostics
+instance NFData   GetHlintDiagnostics
+instance Binary   GetHlintDiagnostics
+
+type instance RuleResult GetHlintDiagnostics = ()
+
+rules :: Rules ()
+rules = do
```

Comment:
I would need `Config` here to enable or disable hlint usage. I guess i could get it using `getClientConfig :: LSP.LspFuncs Config -> IO Config` but i did not find a way to get `LSP.LspFuncs Config` in `Rules a`or `Action a`.
Or maybe should it be disabled in other higher level place? 
//cc @alanz

Reply:
After further investigation:
* It seems `Config` it is not being stored in the shake graph (or at least i have been not able to find it). ghcide seems to ignore it.
  * So it cant be used in `Rules`, the actual way to provide diagnostics in the hls plugin api
* The rest of plugin types (completion, renaming, formatting) are not using `Rules` but `PartialHandler`s, via custom providers so they have access to `LspFuncs Config`
* There is a `DiagnosticProvider` defined (from hie) that could be used to get the config but it is not implemented in hls

So, there would be two alternatives:
* Store the lsp config in the shake graph, creating a Rule for it. We should have in account that it can change at any time by the user, so it have to be changed
* Implement the `DiagnosticProvider` and dont use Rules for provide hlint diagnostics

//cc @alanz

Reply:
I like the idea of a Shake Rule for the config, that way if the user does change the config the diagnostics should then get updated again right?